packages/flutter/lib/src/widgets/implicit_animations.dart

```diff
@@ -565,6 +579,243 @@ abstract class AnimatedWidgetBaseState<T extends ImplicitlyAnimatedWidget> exten
+    this.initialValue,
+    required this.duration,
+    this.curve = Curves.linear,
+    required this.lerp,
+    this.child,
+    this.onEnd,
+  });
+
+  /// Creates an implicit animation using the provided [Duration],
+  /// [ValueWidgetBuilder], and [LerpCallback].
+  ///
+  /// See also:
+  ///
+  ///  * [TweenAnimationBuilder], which does the same,
+  ///    using a [Tween] rather than a [LerpCallback].
+  ///  * [Builder], the [StatelessWidget] equivalent.
+  const factory AnimatedValue.builder(
+    T value, {
+    Key? key,
+    T? initialValue,
+    required Duration duration,
+    Curve curve,
+    required ValueWidgetBuilder<T> builder,
+    required LerpCallback<T> lerp,
+    VoidCallback? onEnd,
+    Widget? child,
+  }) = _AnimatedValueBuilder<T>;
+
+  /// The target value of the animation.
+  ///
+  /// When this value changes, the widget will smoothly animate to the new value
+  /// using the provided [lerp] callback.
+  final T value;
+
+  /// If this is non-null, the widget will immediately start animating from this
+  /// value toward the target [value] when it is first built.
+  final T? initialValue;
+
+  /// The duration over which to animate changes to the [value].
+  final Duration duration;
+
+  /// The curve to apply when the widget animates.
+  ///
+  /// For example, using [Curves.ease] may improve the animation's visual appeal
+  /// by mitigating the linear animation's abrupt stop.
+  final Curve curve;
+
+  /// A function that defines the type [T]'s linear interpolation
+  /// from one [value] to another.
+  ///
+  /// {@macro flutter.animation.LerpCallback}
+  ///
+  /// If the [LerpCallback] is defined as a global function or `static` method,
+  /// it can be used in a `const` constructor!
+  final LerpCallback<T> lerp;
+
+  /// Called each time the animation completes.
+  ///
+  /// This can be useful to trigger additional actions (e.g. another animation)
+  /// at the end of the current animation.
+  final VoidCallback? onEnd;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget? child;
+
+  /// Called each frame to build a widget, given the current state
+  /// of the value's animation.
+  @protected
+  Widget build(BuildContext context, T value);
+
+  @override
+  State<AnimatedValue<T>> createState() => _AnimatedValueState<T>();
+}
+
+class _AnimatedValueState<T extends Object> extends State<AnimatedValue<T>>
+    with SingleTickerProviderStateMixin {
+  late T from = widget.initialValue ?? widget.value;
+  late T target = widget.value;
+  late T value = from;
+
+  Ticker? _ticker;
+
+  void _animate() {
+    final T newTarget = widget.value;
+    if (mounted && newTarget != value) {
+      _ticker?.stop();
+
+      if (widget.duration == Duration.zero) {
+        widget.onEnd?.call();
+        value = newTarget;
+      } else {
+        from = value;
+        target = newTarget;
+        value = widget.lerp(from, target, 0)!;
```

Comment:
So... if lerp would actually return null we would crash? Should we then actually type it as not allowing null?

Reply:
Why do we want to stay compatible with something that would crash here? If we want to stay compatible, we'd somehow handle the null case gracefully.